File: python/Leetcode868BinaryGap/solution.py
class Solution(object):
    def binaryGap(self, N):
        """
        Compute the largest gap between consecutive 1's in the binary representation of a positive
        integer N. If there aren't any consecutve 1's, then return 0. 

        :type N: int
            Positive
        :rtype: int
        """
        if N < 1:
           return 0 

        largest = 0 

        # First find the first 1, guaranteed to find it.

        while N % 2 == 0 and N > 0:
           N //= 2 

        while N > 0:
            N //= 2 # Remove the current 1.
            gap = 1
            while N % 2 == 0 and N > 0:
                N //= 2
                gap += 1            

            # Make sure to test that we didn't just hit the end of the bits.
            if N > 0 and gap > largest: 
                largest = gap
                
        return largest

File: cpp/LeetCode868BinaryGap/solution.cpp
#include "solution.h"

int Solution::binaryGap(int N) {

    // First check the trivial case.
    if(N < 2)
        return 0;

    // Now we are guaranteed to have atleast one binary 1.
    remove_leading_zeros(N);
    int largest_gap = 0, gap = 0;
    while(N > 0) {
       // Use the last gap found.
       if(gap > largest_gap)
            largest_gap = gap;
       gap = remove_gap(N);
    }

    return largest_gap;
}

void remove_leading_zeros(int &N) {

    // Only do something for positive N.
    if (N < 1)
        return;

    while( N % 2 != 1 ) 
        N /= 2;
}

int remove_gap(int &N) {
    int gap = 0;
    do {
        gap++;
        N /= 2;
    } while( N % 2 != 1 && N > 0);
    return gap++; 
}
